### ISSUE TITLE
Expo-Facebook login won't redirect back to Expo Client app when using login with Facebook app

### ISSUE THREAD

**Issue by nnasiruddin** [expo/expo#6459]
### Environment

Expo CLI 3.9.0 environment info:
    System:
      OS: macOS 10.14.5
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 12.13.1 - ~/.nvm/versions/node/v12.13.1/bin/node
      Yarn: 1.19.2 - ~/.yarn/bin/yarn
      npm: 6.12.1 - ~/.nvm/versions/node/v12.13.1/bin/npm
    IDEs:
      Xcode: 11.0/11M337n - /usr/bin/xcodebuild
    npmPackages:
      expo: ^35.0.0 => 35.0.1 
      react: 16.8.3 => 16.8.3 
      react-native: https://github.com/expo/react-native/archive/sdk-35.0.0.tar.gz => 0.59.8 
      react-navigation: ^4.0.10 => 4.0.10 
    npmGlobalPackages:
      expo-cli: 3.9.0

Expo Client on iOS device

### Steps to Reproduce

I recently tried to add login with Facebook to my application. I just copied the logIn function from here
`https://docs.expo.io/versions/v35.0.0/sdk/facebook/`

When I run my app via Expo Client, and execute the logIn function it

1) Redirects me to a page asking if I want to use the Facebook app for login or use phone or email
2) If I use phone or email login works perfectly fine
3) If I select `Use Facebook App` it takes me to the Facebook app, it says `You have previously logged into [MY_APP] and then I press `Continue` a white screen shows up with the `Cancel` button on the top left. It never goes back to my app and never resolves this function `logInWithReadPermissionsAsync`.

### Expected Behavior

It should have gone back to the Expo Client app where my app was running and it should have resolved the promise and returned `(type === 'success') {` so I could use the token and called GraphQL Apis

### Actual Behavior

A white screen shows up with the `Cancel` button on the top left. It never goes back to my app and never resolves this function `logInWithReadPermissionsAsync`.

### Reproducible Demo

```
 async logIn() {
        try {
            const {
                type,
                token,
            } = await Facebook.logInWithReadPermissionsAsync(`MY_APP_ID`, {
                permissions: ['public_profile', 'email'],
            });
            if (type === 'success') {
                // Get the user's name using Facebook's Graph API
                const response = await fetch(`https://graph.facebook.com/me?access_token=${token}`);
                Alert.alert('Logged in!', `Hi ${(await response.json()).name}!`);
            } else {
                // type === 'cancel'
                alert(`Facebook Login Cancel`);
            }
        } catch ({ message }) {
            alert(`Facebook Login Error: ${message}`);
        }
    }

    render() {
        return (
            <Button onPress={this.logIn}><Text>Login</Text></Button>
        )
    }```

<!--
  Please share a project that reproduces the issue.
  There are two ways to do it:

    * Create a new app using https://snack.expo.io/ and try to reproduce the issue in it.
      This is useful if you roughly know where the problem is, or can’t share the real code.

    * Or, copy your app and remove things until you’re left with the minimal reproducible demo.
      This is useful for finding the root cause. You may then optionally create a Snack.

  This is a good guide to creating bug demos: https://stackoverflow.com/help/mcve
  Once you’re done, copy and paste the link to the Snack or a public GitHub repository below.

  (Even if your issue can only be reproduced in a standalone app, please include a Snack
  or GitHub repo we can use to build a standalone app!)
-->

<!--
  What happens if you skip this step?

  Someone will read your bug report, and maybe will be able to help you,
  but it’s unlikely that it will get much attention from the team. Eventually,
  the issue will likely get closed in favor of issues that have reproducible demos.

  Please remember that:

    * Issues without reproducible demos have a very low priority.
    * The person fixing the bug would have to do that anyway. Please be respectful of their time.
    * You might figure out the issues yourself as you work on extracting it.

  Thanks for helping us help you!
-->

**Comment by rodrigoferra** [comment]
I have the exact same problem, I notice that Expo Facebook tries to open the current Facebook app from the phone and then redirects to a website facebook login, then it returns but its not informing facebook app that the login was done. I tried to add behavior: 'web' as an option but it did not work. It worked when I removed my facebook app. Thanks.

**Comment by cruzach** [comment]
This has been reported before, and should be fixed in [SDK 36 (released yesterday)](https://blog.expo.io/expo-sdk-36-is-now-available-b91897b437fe), could you try upgrading and lmk if you still see this behavior?

**Comment by AdelaPetrut** [comment]
@cruzach I tried now with the updated version 3.6 and unfortunately the same behavior @nnasiruddin explained in the description of the bug, it is still reproducible on my side. The Facebook login does never returns the  logInWithReadPermissionsAsync method (after initializeAsync is done). We remain with the white screen shows up with theCancelbutton on the top left. It never goes back to my app and never resolves this.

**Comment by vasilianeugen** [comment]
SDK 36 didn't solve the problem.

**Comment by CloudFol** [comment]
Did they remove the possibility to force "web" behavior? It does not look like my app recognize this after being upgraded to SDK 36, and therefore people now can select themselves, which gives issues.

**Comment by pbeaume** [comment]
I don't know if it is the same problem but I have a strange problem when i run an expo application : It fail with the expo application downloaded from the store, but not with the application downloaded with the expo-cli (even if the version number in both is 2.14.0)

It result in the error "graphrequest can't be used when facebook sdk isn't fully initialized" (probably due to the change in the sdk that force us to call Facebook.initializeAsync in the v36.0 of expo). The problem is that  the application is still in v35.0...

**Comment by islambou** [comment]
> This has been reported before, and should be fixed in [SDK 36 (released yesterday)](https://blog.expo.io/expo-sdk-36-is-now-available-b91897b437fe), could you try upgrading and lmk if you still see this behavior?

i tried , 
i have sdk 36 with expo-facebook 8.0.0 
still have same issue

**Comment by AdelaPetrut** [comment]
> I don't know if it is the same problem but I have a strange problem when i run an expo application : It fail with the expo application downloaded from the store, but not with the application downloaded with the expo-cli (even if the version number in both is 2.14.0)
> 
> It result in the error "graphrequest can't be used when facebook sdk isn't fully initialized" (probably due to the change in the sdk that force us to call Facebook.initializeAsync in the v36.0 of expo). The problem is that the application is still in v35.0...

@pbeaume I also had this problem yesterday -> the call which is mandatory i 36 to be called Facebook.initializeAsync is no available in 35, but it seems it is on older version no longer called. The only solution I found was to upgrade to 36 and call the Facebook.initializeAsync method...

**Comment by islambou** [comment]
> > I don't know if it is the same problem but I have a strange problem when i run an expo application : It fail with the expo application downloaded from the store, but not with the application downloaded with the expo-cli (even if the version number in both is 2.14.0)
> > It result in the error "graphrequest can't be used when facebook sdk isn't fully initialized" (probably due to the change in the sdk that force us to call Facebook.initializeAsync in the v36.0 of expo). The problem is that the application is still in v35.0...
> 
> @pbeaume I also had this problem yesterday -> the call which is mandatory in version 36 to be called is "Facebook.initializeAsync". This method is not available in 35 in order to manually call it and after the 36 release it seems that is no longer called by Expo. The only solution I found was to upgrade to 36 and call the Facebook.initializeAsync method...

actually it's been fixed :D 

make sure to clear npm cache / expo cache and package-lock.json

**Comment by AdelaPetrut** [comment]
@islambou I made a complete reset of the packages (I forgot previously thank you for the hint). But unfortunately, at least on my side the "Expo-Facebook login won't redirect back to Expo Client app when using login with Facebook app" problem it is still not resolved. Did you manage to correctly test this flow?

**Comment by merschermat** [comment]
I was having the same issue on expo, updated to expo sdk 36 and still having the same behavior. Just for testing i created an ios build and uploaded it to Test Flight,  testing with Test flight I give permission on the FB app, and the page redirects me to the my app and resolve the "Facebook.logInWithReadPermissionsAsync" with type = "success". Don't know if that's the case for everyone, but worth testing.

**Comment by cruzach** [comment]
It's not really clear from this thread if there is a problem, or if there is not. 

If you are having this issue, please try testing on a blank `expo init` project, SDK 36

**Comment by AdelaPetrut** [comment]
@cruzach on my side it is still reproducing. 

I created a blank expo init project (SDK 36) -> using minimal (TypeScript) template. 

The same if happening for IOS: 

I select Use Facebook App it takes me to the Facebook app, it says You have previously logged into [MY_APP] and then I press Continue a white screen shows up with the Cancel button on the top left. It never goes back to my app and never resolves this function 'logInWithReadPermissionsAsync`.

  async function onPress() {
    try {
        await Facebook.initializeAsync(facebook_app_id, facebook_app_name);
        const result = await Facebook.logInWithReadPermissionsAsync({
        permissions: ["public_profile", 'email'],
        });
        if (result.type === 'success') {
        } else {
          // user canceled
        }
      } catch ({ message }) {
        console.log(`Facebook Login Error: ${message}`);
}

**Comment by AdelaPetrut** [comment]
@cruzach on my side it is still reproducing.

I created a blank expo init project (SDK 36) -> using minimal (TypeScript) template.
Loading from within EXPO IOS application, is not working.

The same problem mentioned in the issue description is happening for IOS:

I select Use Facebook App it takes me to the Facebook app, it says You have previously logged into [MY_APP] and then I press Continue a white screen shows up with the Cancel button on the top left. It never goes back to my app and never resolves this function 'logInWithReadPermissionsAsync`.

async function onPress() {
try {
await Facebook.initializeAsync(facebook_app_id, facebook_app_name);
const result = await Facebook.logInWithReadPermissionsAsync({
permissions: ["public_profile", 'email'],
});
if (result.type === 'success') {
} else {
// user canceled
}
} catch ({ message }) {
console.log(Facebook Login Error: ${message});
}

**Comment by kogodabra** [comment]
The issue still actual for me as well. I tried all these approaches ^ to find a solution that will works.
My test application for reproduce https://expo.io/@konstantin1/well-travel
Could you prioritize this issue?

**Comment by sharadrb** [comment]
Same here. I created brand new Expo 36 app with facebook integration and its not working. Once redirect to facebook login, it provides option to login with Facebook app but when I chose facebook app then nothing happens.

**Comment by nnasiruddin** [comment]
I created a brand new app, it is still the same. After upgrading to v36

**Comment by nnasiruddin** [comment]
This is my diagnostics,
It does not redirect when using the Expo Client App
Expo CLI 3.11.1 environment info:
    System:
      OS: macOS 10.14.5
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 12.13.1 - ~/.nvm/versions/node/v12.13.1/bin/node
      Yarn: 1.19.2 - ~/.yarn/bin/yarn
      npm: 6.12.1 - ~/.nvm/versions/node/v12.13.1/bin/npm
    IDEs:
      Xcode: /undefined - /usr/bin/xcodebuild
    npmPackages:
      expo: ~36.0.0 => 36.0.0 
      react: ~16.9.0 => 16.9.0 
      react-native: https://github.com/expo/react-native/archive/sdk-36.0.0.tar.gz => 0.61.4 
    npmGlobalPackages:
      expo-cli: 3.11.1

**Comment by nnasiruddin** [comment]
I uploaded the build to Test Flight and it did redirect back to my app fine.

**Comment by heyhujiao** [comment]
Still having the same issue.

Upgraded to Expo SDK 36, `expo-facebook` package: "~8.0.0" but even after I added the new method to initialise the FBSDK `await Facebook.initializeAsync('358071198439996', 'Hawkins')`, I still get the white screen after trying to login via the native FB app. Tested on Expo local app. Did not work.

**Comment by heyhujiao** [comment]
For those whom Facebook Login worked for them in Expo SDK 36, can you show how your _Facebook Login_ settings in the Facebook Developer Portal look like? I suspect there's some configuration required for the FB Client OAuth setup.

**Comment by cruzach** [comment]
@islambou it might help the people here still having this issue if you share your steps to how you got it working. If there's something else that's necessary besides what we have in our documentation, I can add that in

**Comment by nnasiruddin** [comment]
Here are my steps:

I added the following fields to my app.json
facebookScheme
facebookAppId
facebookDisplayName

You can check the expo-facebook page on where to get them from

I also followed the steps to release an app to Testflight and ensured that my bundleId matches what apple has

When I did not have the app on Testflight installed on my phone it did not redirect back

When I downloaded my app from Test Flight and tried the FB login from there (Not the expo client) it worked fine.

**Comment by cruzach** [comment]
Yeah, this issue is only referring to behavior in the Expo Client app, not standalone apps built with `expo build:[ios|android]`

**Comment by nnasiruddin** [comment]
My bad, it does not work on Client for me then...

**Comment by serusko** [comment]
i can confirm as @nnasiruddin wrote, after setting up proper config its working (i had missing facebookScheme now its working perfectly)
thanks for good job with this plugin
you can close it

**Comment by cruzach** [comment]
@serusko you're referring to a standalone app built with `expo build:ios` or `expo build:android`? This issue is referring to behavior  in the Expo Client app

**Comment by serusko** [comment]
- dev mode expo client app: OK
- "test mode" - published app its downloaded from expo server via profile: OK
- test-flight: OK

I am using expo 36.0.1

**Comment by travislang** [comment]
Same issue for me, checked facebook developer settings and the few fields needed for the expo client are all correct.  Works fine except for clicking "use facebook app"

**Comment by andycorm** [comment]
This seems to still be an issue :(

**Comment by artshow** [comment]
Yes, i have the same problem since yesterday evening...

**Comment by artshow** [comment]
Ok its work with iPhone X and Huawei P20 but no old phones :)

**Comment by webmasterdragon** [comment]
Same here with iPhone 11 production mode.

**Comment by artokun** [comment]
Tested on expo "~30.0.2" still same issue.

**Comment by cruzach** [comment]
I've had trouble redirecting back to my app when the bundle ID of my app at https://developers.facebook.com/ was set to my own bundle ID, rather than the Expo Client bundle id - `host.exp.Exponent`. Please make sure you are following the docs. If it's not redirecting back to your app, it's probably trying to find the app associated with the bundle ID you've put in that field. You can confirm this by installing the standalone app via testflight, and then it will probably redirect to that app

**Comment by msojda** [comment]
That was the first thing I've checked @cruzach :-) I ran into this issue yesterday night, tried everything above but unfortunately with no luck.

I haven't tested the app on Testflight yet, I'm far from this stage yet. The issue happens on the Expo Client. I grant permissions and after clicking "Continue" in the Facebook app I'm being stuck on the blank screen with cancel button in the top left corner.

Here's a screenshot of my FB App settings:

![image](https://user-images.githubusercontent.com/1889741/75169298-921e7780-5728-11ea-8a37-328e6aefb6ad.png)

**Comment by jnthnch** [comment]
Having the same issue on my iPhone expo app. This issue happens on my phone, but not on my Macbook iOS simulator. 

I found that it works correctly if I am logged into Facebook on my iPhone's safari browser. As soon as I sign out of the browser (being logged in / logged out to the Facebook app does not affect this), the bug happens again. 

"expo": "^36.0.0"
"expo-facebook": "~8.0.0"
<img width="1435" alt="Screen Shot 2020-02-26 at 12 05 07 PM" src="https://user-images.githubusercontent.com/32749439/75383462-39b6b980-5891-11ea-8739-37181bd8ece2.png">

**Comment by smedel** [comment]
I had the same issue, but running "expo start" and "Press p to toggle production mode. (current mode: development), after pressing "p" it shows "Press p to toggle production mode. (current mode: production)", then the Facebook authentication works, but i still don't get the email of the user (if anyone know how to get it, please tell me :))

"expo": "~36.0.0",
"expo-facebook": "~8.0.0",

I'm also using firebase
"firebase": "^7.8.2",

Hope this can help you. Feel free to tell me if it was the case.

Sébastien.

**Comment by bonafidehan** [comment]
Same issue with my Expo app on prod iOS. The Facebook app doesn't return to the Expo app after authorization. "host.exp.Exponent" is the bundle ID on my app's Facebook settings.

"expo": "^36.0.0"
"expo-facebook": "~8.0.0"

**Comment by maclaine-san** [comment]
I have the same issue, running on my Expo app. Hope this will get fixed soon.

"expo": "~36.0.0"
"expo-facebook": "^8.0.0"

**Comment by vandervillain** [comment]
I am also experiencing the same issue, but I believe it's because facebook's auth flow forces user to go through the facebook app if installed. So:
My app -> browser login screen -> open facebook app -> user clicks continue
app is setup to go back to expo, but not go back to browser.
Workaround:
Login on your device's default browser, in my case, safari. Then it doesn't redirect user to facebook app for login. Would be great if it was smart enough to do either case though.

**Comment by sharadrb** [comment]
To solve this issue, I had to upgrade to Expo version 36 and add initializeAsync()
await Facebook.initializeAsync(<appid>, <appname>)

Then call await Facebook.logInWithReadPermissionsAsync()

**Comment by appseugene** [comment]
This issue still happens to me. I get the white screen mentioned above: initializeAsync -> logInWithReadPermissionsAsync -> facebook app -> continue (allow access) -> white screen (within facebook app) with Cancel top left. Doesnt redirect back to the app. Switching back to Expo Client have to click Cancel again - only then logInWithReadPermissionsAsync  resolved with cancel type.

Even with production mode = true same issue.

expo 36 and expo-facebook v8

**Comment by nadideozturk** [comment]
I'm having this issue with "expo": "^36.0.0", & "expo-facebook": "~8.0.0".
I'm using Expo client.
Reproducing in both development and production mode.
I made sure I have "host.exp.Exponent" as bundle name since the beginning.
Reproducing on iOS device: iPhone 6S.
The same code works on Android just fine.
I was calling Facebook.initializeAsync(appId) since the beginning. Specifying appName didn't make a difference.
Browser is not logged in to Facebook, since my users don't have to do that.
PS: Tried logging in to Facebook via browser before, now it is not reproducible (until I clear browser cache I think). The point is, it will still fail on users' devices so nothing is fixed.

**Comment by andycorm** [comment]
Reading @cruzach's comment helped me solve this issue 🎉

For the call to `Facebook.initializeAsync` I was passing in my live bundle ID, not the expo one (`host.exp.exponent`). I changed my code like-so and now it's working great: (expo version 36)

```
import Constants from 'expo-constants';

const APP_ID = '1234123412341234'; // Facebook App ID here
const BUNDLE_ID = Constants.appOwnership === 'expo' ? 'host.exp.Exponent' : 'com.my_real.app_id';
await Facebook.initializeAsync(APP_ID, BUNDLE_ID);
```

**Comment by appseugene** [comment]
@andycorm , hmm it still doesnt work for me. By default in Expo Client it should choose 'host.exp.Exponent' as the bundle when the second argument to initializeAsync is not provided. But i did do like you said and still no redirect back to the expo client.
Do yo have both 'host.exp.Exponent' and 'com.my_real.app_id' listed on Facebook's iOS settings in dashboard?

**Comment by ceszgdav** [comment]
actually  I was having the same problem but in the documentation says you need to initialize the SDK before the  `logInWithReadPermissionsAsync` so I wrote my code like this...

`const login = async () => {
        Facebook.initializeAsync(FacebookApi.application_id);
        const { type, token } = await Facebook.logInWithReadPermissionsAsync(
            { permissions: FacebookApi.permissions }
        );
        console.log(type);
    };`

and that way works for me.

I hope is helpful for someone

**Comment by appseugene** [comment]
ceszgdav pls read the descriptions/title above. what you mentioned  is a separate issue with Expo Client update and SDK 36, where you don't even get to the login. This is AFTER you are logged into Facebook, it won't redirect back with success type.

**Comment by harrymelka** [comment]
To fix this issue you just have to modify the bundle ID with the real one in the facebook developers page:
Instead of host.exp.Exponent you put your com.yourapp and it should work.

**Comment by appseugene** [comment]
I've tried that and no effect at all. Same issue, once I hit Continue in Facebook App, the screen goes white and only a top left Cancel button remains on screen. Nothing happens afterwards.

**Comment by thomasgit123** [comment]
Same issue for me with latest version

**Comment by bonafidehan** [comment]
@andycorm @appseugene Can also confirm, tried to explicitly specify 'host.exp.Exponent' when calling initializeAsync. Did not work, same problem.

**Comment by Harrison97** [comment]
What are you guys doing to get around this? I have the same issues with latest versions.

**Comment by vandervillain** [comment]
I can make it happen 100% of the time on my iPhone 6S. replication steps:
1. expo init testFacebook  (configure as managed workflow)
2. expo install expo-facebook
3. should have these now in dependencies:
    "expo": "~36.0.0",
    "expo-facebook": "~8.0.0"
4. in my developer.facebook app, under IOS, set Bundle ID to host.exp.Exponent
5. add this to my App.tsx:
```
async componentDidMount() {
  await Facebook.initializeAsync([YOUR_APP_ID], [YOUR_APP_NAME])
  const result = await Facebook.logInWithReadPermissionsAsync({ permissions: ['public_profile', 'email'] })
  if (result.type === 'success') {
    console.log('success')
  } else {
    // user canceled
  }
}
```
6. install facebook app if not already installed
7. open safari, go to facebook.com, logout, make sure to click the gear next to your profile and "Remove account from device". This will make it seem like you have yet to login to fb
8. start/open your app, "Expo Wants to Use facebook.com..." -> continue 
9. "Open this page in "Facebook"? -> Open 
10. for me, "You previously logged in to APP_NAME with Facebook" (I have logged in successfully previously when safari was logged in) -> Continue 

you'll end up on an empty white screen with "cancel" button in top left.
If you feel like it, open safari, facebook.com, login, and then try again. You'll be able to get into your app just fine. Obviously won't be sufficient for an actual production application.

**Comment by bonafidehan** [comment]
@cruzach is there a way we can provide more help / data to narrow down the bug so that it can be fixed? cc @appseugene & @vandervillain

**Comment by cruzach** [comment]
@vandervillain following your exact steps worked for me in the expo Client. After installing the Facebook app, did you log in to the app? (i did not)

Sorry everyone, it's obvious that there is some issue here but without a reliable way to reproduce it it's very hard to debug on our end 🙁 If you want to investigate in your own further, I think the best next step is to take a look at the device logs and see if there's anything interesting there, and then I'd also recommend testing `expo-facebook` in a bare workflow project

**Comment by vandervillain** [comment]
@cruzach I encounter the issue whether I am logged in to the app or not--I will continue to investigate on my end. Thanks!

**Comment by appseugene** [comment]
@cruzach , it works fine for you on sdk 36 with expo-facebook 8.0.0 ? while logged into facebook app (and not in safari), does it redirect back to the Expo Client?
if so, pls share your configations in Facebook Developer Console also

**Comment by gitteraz** [comment]
I Can confirm every non working examples.

**Working example**: While logged in to Facebook via Safari Browser, initiate FB Login from expo client App. The login process will happen in the browser and redirect back to expo client app.

For me the redirection does not happen when Login happens in the Facebook app and not in the browser.

Expo SDK 36.0.0
expo-facebook 8.0.0

**Comment by cruzach** [comment]
@appseugene 
<img width="970" alt="Screen Shot 2020-03-26 at 10 11 45 AM" src="https://user-images.githubusercontent.com/35579283/77656472-4f7bd500-6f4a-11ea-86b4-045225a0e17d.png">
<img width="966" alt="Screen Shot 2020-03-26 at 10 12 12 AM" src="https://user-images.githubusercontent.com/35579283/77656473-50ad0200-6f4a-11ea-8b1f-566a7631e4e9.png">

**Comment by bonafidehan** [comment]
The only deltas I could detect from your FB settings and mine were:

- "Valid OAuth Redirect URIs": I have a Firebase URL there
- And likely a delta, my FB app is "In development"

@cruzach can you check what version of Facebook iOS app you have? Mine's 262.0.0.47.113.

cc @appseugene

**Comment by asabadyr** [comment]
> The only deltas I could detect from your FB settings and mine were:
> 
> * "Valid OAuth Redirect URIs": I have a Firebase URL there
> * And likely a delta, my FB app is "In development"
> 
> @cruzach can you check what version of Facebook iOS app you have? Mine's 262.0.0.47.113.
> 
> cc @appseugene

We experienced the same trouble with iOS build in TestFlight.
One person was all the time redirected to the FB app to authorise and then grown back in our app but with no login status. 
I've asked to remove both our app and FB app. Then install FB app and our app back. 
Authorization through Facebook App worked, and our app got login status.
Could it be a problem with outdated FB app? May be, but person has auto install updates from AppStore enabled.
Looks really weird and hard to trace. And what is even worse we now need to ask people to re-install FB app in order to start using our app. That definitely a very bad experience.

**Comment by bonafidehan** [comment]
@cruzach any chance I could just jump on a video call with you and figure this out? The alternative is to disable FB login for iOS and that would really suck for a lot of apps. Would love to spend any time needed to unblock you on this. By myself, it's hard to figure out what other info to provide. The call would be super helpful.

**Comment by AndreiChe** [comment]
@cruzach Please correct me if I’m wrong. The expo-facebook 8.0.0 uses facebook-sdk 5.8 under the hood. The “behavior” option was deprecated from 4.43.0. That’s why the browser behavior is the only option. We use expo for a long time and previously used browser behavior only (native app was always buggy) and users didn’t have an option to launch the application from the in-app browser. Now they can.. Might it be the core of the issue?

**Comment by flo-butterfly** [comment]
Meanwhile is it possible to force browser identification instead of opening the app as the current behavior (white screen) is bad for the user experience and can be misleading.

Thank you

**Comment by maclaine-san** [comment]
Anyone tried expo SDK 37 for this issue?

**Comment by sharadrb** [comment]
By the way, we did following and issue was resolve
- Upgraded to Expo version 36
We just didn't upgrade. We had to recreate new project and copied all project files. We had to create new app because we upgraded all our components including react navigation.

- await Facebook.initializeAsync("appid", "appname")
- call await Facebook.logInWithReadPermissionsAsync()

Try creating brand new expo app and try to call initializeAsync() followed by logInWithReadPermissionsAsync()

Hope this helps

**Comment by FatBuTa** [comment]
Hi all,
Please try to use yarn instead of npm
Hope it will help.

**Comment by bonafidehan** [comment]
@maclaine-san verified exact same problem exists with Expo SDK 37:

- "expo": "^37.0.0"
- "expo-facebook": "~8.1.0"

@sharadrb Yup that's exactly what we're all doing and still encountering problems.
```
  await Facebook.initializeAsync(APP_ID, 'host.exp.Exponent')
  const {
    type,
    token,
    expires,
    permissions,
    declinedPermissions,
  } = await Facebook.logInWithReadPermissionsAsync({
    permissions: ['public_profile', 'email'],
  })
```

**Comment by bonafidehan** [comment]
@AndreiChe your hypothesis in https://github.com/expo/expo/issues/6459#issuecomment-606544783 seems possible. But would you be able to reason why it impacts (apparently) only some use cases? For example, cruzach is not able to reproduce the bug.

**Comment by FatBuTa** [comment]
As I investigate some use cases:
Create a new expo project with Facebook login only.
1. try to create a new facebook project.
==> after login it display the blank screen with Cancel button
2. try to use existing facebook project that worked for a long time.
==> after login it open the app that registered in the facebook project.
3. try to modify the existing facebook project above to point the Bundle ID to host.exp.Exponent
==> after login it open the app that registered in the facebook project.(exactly same with 2)
So I think the problem is facebook developer page have a problem with update Bundle ID

**Comment by AndreiChe** [comment]
@bonafidehan To be honest, I don't know exactly what is the issue. We tried many suggested fixes including creating a new app but nothing worked stably for us. This is just one more idea. I guess that if Facebook says that they support the browser behavior only, the authentication using the app will not work properly. @cruzach please contact me if you need help with reproducing the issue.

**Comment by cruzach** [comment]
@AndreiChe I was able to reproduce this in the Expo client app, can you confirm for me that this works as expected in a standalone app built with `expo build:ios`?

**Comment by AndreiChe** [comment]
@cruzach No, it doesn't. Of course, we didn’t do a release in stores but tried all these fixes on the ios test flight. It wasn’t stable nor in the Expo client neither in the standalone builds

**Comment by cruzach** [comment]
Same case that reproduces it in the Expo Client is working fine for me in testflight on SDK 37

**Comment by AndreiChe** [comment]
@cruzach Please clarify a bit for my understanding. As I mentioned above, when users are on the Facebook page in in-app browser they see a popup that allows them to authenticate via the app. You mean they don't have such an option on SDK 37 anymore. Or was it another issue?

**Comment by cruzach** [comment]
@AndreiChe in a standalone SDK 37 app, I was not given the option to authenticate via the Facebook app.  This is the login function I used:
```
async function logIn() {
  try {
    await Facebook.initializeAsync("MY-FB-APP-ID");
    const {
      type,
      token
    } = await Facebook.logInWithReadPermissionsAsync({
      permissions: ["public_profile"]
    });
    if (type === "success") {
      // Get the user's name using Facebook's Graph API
      const response = await fetch(
        `https://graph.facebook.com/me?access_token=${token}`
      );
      Alert.alert("Logged in!", `Hi ${(await response.json()).name}!`);
    } else {
      // type === 'cancel'
    }
  } catch ({ message }) {
    alert(`Facebook Login Error: ${message}`);
  }
}
```

In the facebok developer console, in the app's advanced settings, I have 
- `Native or desktop app? -Yes` 
- SDK version: iOSv5.8.0

And for Facebook Login settings, I have:
<img width="815" alt="Screen Shot 2020-04-03 at 1 58 44 PM" src="https://user-images.githubusercontent.com/35579283/78390878-50001580-75b3-11ea-8272-3daf79c22f4c.png">

**Comment by gvillenave** [comment]
Hi everyone, has there been any progress on this issue? I'm hitting this on a brand new empty Expo Client app on SDK 37.

**Comment by YahiaBadr** [comment]
@bonafidehan did you find any work around for Expo sdk 37 and expo-faceboo 8.1.0 ?

**Comment by bonafidehan** [comment]
@YahiaBadr I didn't find a workaround. Current status that I'm aware of: cruzach was able to reproduce for Expo client app but not for standalone.

**Comment by YahiaBadr** [comment]
@bonafidehan  It still doesn't work for me in the Expo client app, if i use safari for auth it redirects. However, if i use the Facebook app it sticks in the blank page with the cancel button. Did it work for you in the Expo client app with the Facebook app?

**Comment by bonafidehan** [comment]
@YahiaBadr please read my and other comments more carefully. As I already mentioned, cruzach has reproduced the bug in Expo client (https://github.com/expo/expo/issues/6459#issuecomment-608441099) and is likely working on a fix.

**Comment by alikhaledahmed** [comment]
SOLUTION(IOS): Remove Facebook app from your phone and try doing it with Safari browser (worked fine for me)

**Comment by YahiaBadr** [comment]
> SOLUTION(IOS): Remove Facebook app from your phone and try doing it with Safari browser (worked fine for me)

This solution is only for you as a developer, but we cannot publish our apps like this. I had to remove the Facebook login from my app when i publish it as it will be very inconvenient for the user.

**Comment by ArthurDelbeke** [comment]
Hi, 
I've also the same problem only on my device, with the simulator it's ok, but on real device, white screen like said before...

I'm under the 37.0.0 SDK,

Don't want to use this solution because as @YahiaBadr said, we cannot publish our apps like that...

> SOLUTION(IOS): Remove Facebook app from your phone and try doing it with Safari browser (worked fine for me)

**Comment by AndreiChe** [comment]
Hi @cruzach. Thank you for the detailed instructions. We have upgraded the expo version and updated the configuration on the Facebook developer's page. Now we have exactly the same setup as you have. This is how we login:

`
await Facebook.initializeAsync(secrets.facebook.appId);
const result = await Facebook.logInWithReadPermissionsAsync(Config.FACEBOOK.OPTIONS);
`

Issues:
- Users still have an option to launch the Facebook app
- This issue is still here. It is not reproduced consistently. Sometimes happens on terminating the app and first login attempt; sometimes after several attempts to log in / log out.

**Comment by AndreiChe** [comment]
Hi @cruzach. We have ejected the app to manage the Facebook SDK directly. It worked exactly the same way as with the Expo, so that users had an option to launch the Fecabook app. Then, after doing this fix https://github.com/expo/expo/issues/6112#issuecomment-610546525 we got browser behavior only! So the issue has been resolved now. Unfortunately, we didn’t find a way to make Facebook app work but it seems to be impossible for now, which is sad from the UX perspective..
This could be also related: https://github.com/facebook/facebook-ios-sdk/issues/1067#issuecomment-578875814

**Comment by cruzach** [comment]
The Facebook app login workflow is no longer supported, I believe. Peak at the changes to `FBSDKLoginBehavior` between [4.39](https://developers.facebook.com/docs/reference/iossdk/4.39.1/FBSDKLoginKit/enums/fbsdkloginbehavior.html/) and [v5](https://developers.facebook.com/docs/reference/iossdk/5.0.0/FBSDKLoginKit/enums/fbsdkloginbehavior.html/)

**Comment by asabadyr** [comment]
@cruzach any ideas what should we all do? If we stay in managed flow. It's a real pain for us

**Comment by YahiaBadr** [comment]
any updates?

**Comment by asabadyr** [comment]
I've created a workaround for this situation.
I put a timeout for the authentication promise and use a fallback flow via [AuthSession](https://docs.expo.io/versions/v36.0.0/sdk/auth-session/)

**Comment by YahiaBadr** [comment]
@asabadyr could you share the code with us?

**Comment by dpmika** [comment]
Hi, 
i have same problem, @asabadyr  could you share the code with us?

**Comment by dpmika** [comment]
Hi guys, 
I confirm that the problem only appears when the application is run on the device via the expo client. 
Today I did a test by doing a build and sending it with TestFlight. 
The login works !

**Comment by webmasterdragon** [comment]
Any update which can be resolved this issue?

**Comment by dpmika** [comment]
@webmasterdragon No, but the problem is only there if you run the app on Expo client.  On native apps, there's no problem

**Comment by asabadyr** [comment]
> @webmasterdragon No, but the problem is only there if you run the app on Expo client. On native apps, there's no problem

incorrect. we still face this from time to time without any logic

**Comment by asabadyr** [comment]
> Hi,
> i have same problem, @asabadyr could you share the code with us?

`        promiseTimeout(30000, promise).then(({
                                               type,
                                               token,
                                               expires,
                                               permissions,
                                               declinedPermissions,
                                           }) => {
            if (type === 'success') {
                const credential = firebase.auth.FacebookAuthProvider.credential(token);
                firebase.auth().signInWithCredential(credential)
                    .then(v => {
                        ...
                    })
                    .catch((error) => {
                        alert(lstr('facebookLoginError', {error: error}));
                    })
            } else {
                alert(lstr('facebookLoginError', {error: type}));
            }
        }).catch(err => {
            setUseAltFlow(true);
        });
`

`    const login2 = async () => {
        setButtonDisabled(true);
        let redirectUrl = AuthSession.getRedirectUrl();
        console.log(redirectUrl);
        let result = await AuthSession.startAsync(
            {
                authUrl:
                    `https://www.facebook.com/v2.8/dialog/oauth?response_type=token`  +
                    `&client_id=${Config.loginSetup.facebookID}` +
                    `&redirect_uri=${
                        encodeURIComponent(
                            redirectUrl
                        ) }`,
            }
        );
}
`

**Comment by vineetvk01** [comment]
I am facing the same issue. 
After Login once from my account. whenever i try to re-authenticate the user. It doesn't ask for username and password screen again.

**Comment by cruzach** [comment]
We'll be releasing an update to the App Store Expo Client with this fix included soon, most likely next week

**Comment by LucasMilotich** [comment]
hello, is there any update of this issue? thanks!

**Comment by cruzach** [comment]
Fix has landed, you must update to the most recent version (2.15.5) of the Expo Client app to use it

**Comment by Amirjab21** [comment]
Is the fix specific to the simulator or does it fix issues regarding actual device experience?

**Comment by AdelaPetrut** [comment]
@cruzach is the fix available also for expo bare application? using EXPO 37 version?

**Comment by cruzach** [comment]
@Amirjab21 the fix is for both sims and devices

@AdelaPetrut this issue didn't affect bare apps afaik

**Comment by AdelaPetrut** [comment]
@cruzach unfortunately we have a bare app which was reproducing the:
_[iOS][Facebook] Not redirecting back to app after Facebook.logInWithReadPermissionsAsync through facebook_

Related I think also to the following GIT (isn't the correction also the same for the below GIT issue)?:
https://github.com/expo/expo/issues/6112

**Comment by cruzach** [comment]
This issue is for the App Store Expo Client app, not standalone or bare apps. If you're seeing this issue in a bare app please create a new issue with a reproducible demo included.

Most likely you have not included the relevant facebookScheme and facebookAppId in info.plist if you're seeing this in the bare workflow

**Comment by AdelaPetrut** [comment]
> This issue is for the App Store Expo Client app, not standalone or bare apps. If you're seeing this issue in a bare app please create a new issue with a reproducible demo included.
> 
> Most likely you have not included the relevant facebookScheme and facebookAppId in info.plist if you're seeing this in the bare workflow

Thank you for the details.

We are using the ExpoKit - and unfortunately updating to EXPO 37 does not solve the issue :(

**Comment by amnayerji** [comment]
Hi, I am using expo sdk 40 and expo-auth-session 3.0.0 and I am having this issue. Any fixes?

**Comment by CaioMars** [comment]
Same here @amnayerji, did you find a solution?

**Comment by Muhammad-Saleet** [comment]
> Same here @amnayerji, did you find a solution?

I've solved this problem by using [expo-facebook](https://docs.expo.io/versions/latest/sdk/facebook/) instead of expo-auth-session. 

The code ended up looking something like this:

```
  async function loginWithFacebook () {
    console.log('logging in with facebook')

    await Facebook.initializeAsync({
      appId: Keys.FACEBOOK_APP_ID,
      appName: Keys.FACEBOOK_APP_NAME,
    })
    const { type, token } = await Facebook.logInWithReadPermissionsAsync({
      permissions: ['public_profile', 'email'],
    })

    if (type !== 'success') {
      console.log('token: ', token)
    } else {
      console.log('login failed or cancelled')
    }
  }
```

**Comment by CaioMars** [comment]
Thanks for the quick response @muhajirdev! You're the man! I got it to work with `expo-facebook`, it's such a relief! Thank you!!!

**Comment by muhajirdev** [comment]
haha,
I think you should mentioned wrong person @CaioMars

**Comment by CaioMars** [comment]
Hahaha! Yeah, i did, I meant @Muhammad-Saleet... Thanks!

**Comment by dovelive** [comment]
@Muhammad-Saleet @CaioMars 


```
  import * as Facebook from 'expo-facebook'

  async function loginWithFacebook () {
    console.log('logging in with facebook')

    await Facebook.initializeAsync({
      appId: Keys.FACEBOOK_APP_ID,
      appName: Keys.FACEBOOK_APP_NAME,
    })
    const { type, token } = await Facebook.logInWithReadPermissionsAsync({
      permissions: ['public_profile', 'email'],
    })

    if (type !== 'success') {
      console.log('token: ', token)
    } else {
      console.log('login failed or cancelled')
    }
  }
```

I used that code, but error is still occurred.
Do you really think that that works well?

**Comment by CaioMars** [comment]
@dovelive it looks good, exactly what I have. What error are you getting? Are you running it on Expo or on a build?

**Comment by dovelive** [comment]
@CaioMars 
Thank you for contacting.
Error occurs on the build environment. But it is in bare workflow.
And the error that I am getting is - when I am going to login with facebook app, it isn't redirected to the app, so I can't get the result from Facebook.logInWithReadPermissionsAsync function.
RNApp -> Builtin Webview -> Facebook App -> Builtin Webview -> Facebook App ...

**Comment by CaioMars** [comment]
I'm using Bare workflow too. It sounds like it's a configuration problem to me, your code looks good. There are a lot of little steps to configure it properly. Did you follow the instructions in [Expo Facebook Auth Documentation](https://docs.expo.dev/versions/latest/sdk/facebook/)?

Full disclosure, I am no expert on third-party authentication. It's one of those things I rarely do and by the time I have to do it again I already forgot how to. But here are few things to check:

**Facebook Dev console  (App > Settings > Basic):**

For iOS:
- Did you add bundleId in the Facebook Dev console > App > Settings > Basic (iOS)?

For Android:
- Did you add Package name (same as bundleId)?
- Did you add Class Name? Usually same as bundleId with your main class name from your Android Manifest (ex: com.my.app.MainActivity).
- Did you add your Key Hashes? 

To see your key hashes run in the root of your project:

    expo fetch:android:hashes

**app.json**

Make sure you have added these values to your `app.json` file. Note that one has `fb` preceeding your app ID and the other doesn't.

    {
        "expo": {
            ...
            "facebookScheme": "fb{FACEBOOK_APP_ID}",
            "facebookAppId": "{FACEBOOK_APP_ID}",
            "facebookDisplayName": "{APP_NAME}"
        }
    }

**Android Manifest**

Add to your manifest:

    <data android:scheme="fb{FACEBOOK_APP_ID}"/>

I hope this helps.

**Comment by dovelive** [comment]
For android, it works well.
Error is only occurred on iOS.
Once user turns into expo bare workflow, app.json is no longer used.
However thank you for your comment, and I will re-check my settings.

**Comment by CaioMars** [comment]
I see. Did you add this to your `info.plist`?

    <dict>
        <key>CFBundleURLSchemes</key>
        <array>
        <string>fb{FACEBOOK_APP_ID}</string>
        </array>
    </dict>

**Comment by dovelive** [comment]
Yes. I think it was automatically configured when turned into bare workflow.

**Comment by seongwoobyun** [comment]
I'm seeing this issue with SDK 43 when i build a standalone app via EAS. Anyone have the same issue?

**Comment by nurimanizam** [comment]
> I'm seeing this issue with SDK 43 when i build a standalone app via EAS (managed workflow). Anyone have the same issue?

 I have an app on SDK42 using EAS and Facebook Login doesn't work for Android Build but for Classic build, it's working fine.

**Comment by CaioMars** [comment]
@nurimanizam, did you figure it out?

My is working but I often have problems with it. What kind of error are you getting?

**Comment by nurimanizam** [comment]
@CaioMars I went back to expo build instead of eas build for the time being. There is probably a bug with the eas build at that time of building.

**Comment by lorenzo-ange** [comment]
Same for me, building with expo build instead of eas build resolved the issue. :(
